mlflow/evaluation.py

```diff
@@ -0,0 +1,354 @@
+
+
+class ModelEvaluator:
+
+    def can_evaluate(
+        self, model_type, evaluator_config=None, **kwargs
+    ) -> bool:
+        """
+        :param model_type: A string describing the model type (e.g., "regressor",
+                           "classifier", …).
+        :param evaluator_config: A dictionary of additional configurations for
+                                 the evaluator.
+        :param **kwargs: For forwards compatibility, a placeholder for additional
+                         arguments that may be added to the evaluation interface
+                         in the future.
+        :return: True if the evaluator can evaluate the specified model on the
+                 specified dataset. False otherwise.
+        """
+        raise NotImplementedError()
+
+    def compute_metrics(self, predict, dataset):
+        """
+        return an instance of EvaluationMetrics
+        """
+        raise NotImplementedError()
+
+    def compute_and_log_artifacts(self, predict, dataset, run_id, mlflow_client):
+        """
+        compute and log artifact, and return a dict of
+        artifact_name -> instance_of_EvaluationArtifacts
+        """
+        raise NotImplementedError()
+
+    def evaluate(
+        self, predict, dataset, run_id=None, evaluator_config=None, **kwargs
+    ) -> EvaluationResult:
+        """
+        :param predict: A function used to compute model predictions. Predict
+                        accepts features from the specified `dataset` and
+                        feeds them to the model, producing output predictions.
+        :param dataset: An instance of `EvaluationDataset` containing features
+                        and labels (optional) for model evaluation.
+        :param run_id: The ID of the MLflow Run to which to log results.
+        :param evaluator_config: A dictionary of additional configurations for
+                                 the evaluator.
+        :param **kwargs: For forwards compatibility, a placeholder for additional
+                         arguments that may be added to the evaluation interface
+                         in the future.
+        :return: An `EvaluationResult` instance containing evaluation results.
+        """
+        client = mlflow.tracking.MlflowClient()
+        with GetOrCreateRunId(run_id) as run_id:
+            metrics_dict = self.compute_metrics(predict, dataset)
+            timestamp = int(time.time() * 1000)
+            dataset_id = dataset.name if dataset.name is not None else dataset.hash
+            # TODO: log tags of dataset metadata
+            client.log_batch(
+                run_id,
+                metrics=[
+                    Metric(key=f'{key}_on_{dataset_id}', value=value, timestamp=timestamp, step=0)
+                    for key, value in metrics_dict
+                ],
+            )
+            artifact_dict = self.compute_and_log_artifact(predict, dataset, run_id, client)
```

Comment:
When log artifact, the artifact path shall we also include the dataset name/hash ?

Reply:
we should definitely include the dataset name. I don't think we need to include the hash.